### ISSUE TITLE
Implement cache revalidation

### ISSUE THREAD

**Issue by shagamemnon** [cloudflare/kv-asset-handler#62]
Today, cacheable, unchanged resources are not stored in the browser cache - regardless of the Cache-Control response headers that are applied in Workers. In other words, it's impossible to get a 304 Not Modified response (it is, however, possible to get Chrome to return the resource from memory cache).

There are two reasons for this:
1. Resources stored in Cache API do not include a `Last-Modified` or `Etag` response header
2. The cacheKey tied to each resource would strip the `If-Modified-Since` or `If-None-Match` header before querying the cache API (right now, the browser wouldn't even send these headers because of the first reason)

**Solutions**
Coming soon

**Comment by victoriabernard92** [comment]
Thanks for this. One can set the browser TTL in the `getAssetFromKV` handler. See: https://github.com/cloudflare/kv-asset-handler#browserttl.  I spent a big lot of testing out our cache API today for the more optimized caching as you suggested. What I came up with was:

Since the cache API's [`match`](https://developers.cloudflare.com/workers/reference/apis/cache/#match) does respect `If-Modified-Since` , `Etag` and `If-None-Match`, the edge should respond appropriately for browser's sending those headers. 

Behavior expected from an origin though - e.g as you mentioned returning 304s when appropriate - are not implemented. But one can implement this by modifying the internals of `workers-site`. Please let us know if you come up with a solution you'd like to share and I'd be happy to review.

**Comment by budparr** [comment]
I was having a caching issue with CF Workers. I posted about it in the forums with my solution. Not sure if it's helpful to you or not, but I'll share here. https://community.cloudflare.com/t/workers-sites-changes-not-reflected-in-browser-until-hard-refresh/141378/4?u=ops3

**Comment by shagamemnon** [comment]
Thanks @budparr! Appreciate the input.

I do think that the `options` parameter should support `etag` generation. As @victoriabernard92 mentioned (and you observed) `cache.match()` already supports conditional revalidation, permitted an `etag` is provided by an origin server.

A solution here needs to consider the following:
- There needs to be a collision-resistant digest created for every version of file (e.g. the `etag`)
- On cache insertion, two objects will need to be created -- the 304 Not Modified response and the 200 OK response with the cache body
- We need to choose whether the etags should be strong or weak
- Resources will need to have a different etag for each content-encoding, assuming strong etags are used

 Here is my plan:
- Workers supports the WebCrypto API, which includes a `digest()` method that can be used to hash the actual content bodies. So we can kill two birds with one stone here by hashing the response body contents upon cache insertion like `digest('SHA-1', '<html>...')`. This ensures a new version of a file will always have a different `etag` than its predecessor
- The etag will need to be dropped in scenarios where the size of the file being inserted into cache is too large for Workers to compute in the 5ms CPU limit. I'm estimating that this limit will be somewhere between 2MB and 6MB
- If an etag is sent from an origin server, a new etag will not be generated in the Worker
- Strong e-tags will be used for two reasons:  1) Cache API allows (and encourages) the creation of unique objects that are derived (but not identical to) the origin response and 3) we want to support Range-Requests. Range Requests aren't fulfilled when the etag is weak
- Each content tag will be in the format `SHA1(HASH-OF-BODY)-{content-encoding}` such that content-encoding differences are handled properly

**Comment by EverlastingBugstopper** [comment]
@shagamemnon - this sounds good. fwiw we already hash all of the files served by a workers site and are included in the key in KV so we may be able to just reuse that.

**Comment by EatonZ** [comment]
Following this as well. It's been a few months, so have there been any updates on this feature?

I have been working on caching in my Workers app and it would be nice to have the ETag situation straightened out. At this point, it's looking like I need to generate an ETag myself, and then check it later by comparing it against If-None-Match. It seems like this should be built-in somehow instead of me having to code in a solution.

**Comment by shagamemnon** [comment]
Hey @EatonZ - thanks for flagging this. It got de-prioritized over the last month or so.

I have a preliminary solution complete but need to talk to @EverlastingBugstopper because it does not make use of the wrangler-generated manifest. Should be able to make a PR by end of week. With respect to your comment about CPU: of course, that are the only considerations here :)

**Comment by EatonZ** [comment]
Hi @shagamemnon @EverlastingBugstopper, let me know when you have something ready - happy to try it out.

**Comment by whatnickcodes** [comment]
Assuming this PR gets merged, what do you all now recommend for `edgettl` and `browserttl` now?

Like, what's a reasonable / recommended balance between super blazing fast and also having request clear in timely fashion?

Any opinions / thoughts?

**Comment by whatnickcodes** [comment]
Following up with results from my own tests, if ETag's do get merged, I think you actually should be able to set both **browserTTL and edgeTTL to a full year**!

Just so long deployment has a cache clear to Cloudflare's edges, ETag keys relate to a file hash, and you are truly static.

For the default KV-Asset Handler library I don't think it makes sense to have this setting be the default since lot of people use this library beyond simple Worker Sites.

But for something that is 100% static with nothing dynamic, this would be perfect to get blazing speeds from this pull request.

Total noob at this but I think we have:

**First Route Visit**
_Client Request_ asset for first time (html/css/img/pdf/js/whatever)
_Server Responds_ 200 OK (head and body)

**Returning Route Visit**
_Client Request_ asset for second time (same file) with ETag
_Server  Responds_ 304 with ETag match from Edge (head only)

**Returning Route Visit (with a new edge cache from a clear)**
_Client Request_ asset for third time (same file) with ETag
_Server  Responds_ with ETag mismatch regardless of their browserTTL, 200 (head and body) sent, process starts over

**Comment by shagamemnon** [comment]
Hey @whatnickcodes - your first statement is correct. edgeTTL and browserTTL could have a `max-age=infinity;` etags are one of many techniques to handle cache update and validation.

One of the reasons we want to implement it and intend to make it default behavior is that etags live harmoniously with dynamic content. This is because a) etags guarantee the integrity of a resource and b) are not specific to cache. For example, etags are actually employed in a different context by PUT requests to aid in concurrency control (see https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests#Avoiding_the_lost_update_problem_with_optimistic_locking)

The TL;DR here is that adding etags won't negatively affect request for dynamic resources

**Comment by simplenotezy** [comment]
Can't wait for this PR

**Comment by theromis** [comment]
me too :)
